# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: js/src/vm/HelperThreads.cpp
# Commit: 68e6b05b17b7
# Full Hash: 68e6b05b17b7e32496b5519f0b0fd9e99e10b3a0
# Author: Jon Coppeard <jcoppeard@mozilla.com>
# Date: 2020-10-28 09:24:21
# Description:
#   Bug 1672760 - Add an API to wait for parse tasks to complete without cancelling them r=sfink
#   
#   The problem here was cancelling parse tasks without the browser's knowledge (I
#   didn't realise that the cancel method did anything beyond waiting).
#   
# ==============================================================================

diff -r b55c94de3ec7 -r 68e6b05b17b7 js/src/vm/HelperThreads.cpp
--- a/js/src/vm/HelperThreads.cpp	Tue Oct 27 15:24:54 2020 +0000
+++ b/js/src/vm/HelperThreads.cpp	Tue Oct 27 14:10:27 2020 +0000
@@ -879,24 +879,12 @@
   }
 }
 
-void js::CancelOffThreadParses(JSRuntime* rt) {
-  AutoLockHelperThreadState lock;
-
+static void WaitForOffThreadParses(JSRuntime* rt,
+                                   AutoLockHelperThreadState& lock) {
   if (HelperThreadState().threads(lock).empty()) {
     return;
   }
 
-#ifdef DEBUG
-  GlobalHelperThreadState::ParseTaskVector& waitingOnGC =
-      HelperThreadState().parseWaitingOnGC(lock);
-  for (size_t i = 0; i < waitingOnGC.length(); i++) {
-    MOZ_ASSERT(!waitingOnGC[i]->runtimeMatches(rt));
-  }
-#endif
-
-  // Instead of forcibly canceling pending parse tasks, just wait for all
-  // scheduled and in progress ones to complete. Otherwise the final GC may not
-  // collect everything due to zones being used off thread.
   while (true) {
     bool pending = false;
     GlobalHelperThreadState::ParseTaskVector& worklist =
@@ -923,6 +911,26 @@
     }
     HelperThreadState().wait(lock, GlobalHelperThreadState::CONSUMER);
   }
+}
+
+void js::WaitForOffThreadParses(JSRuntime* rt) {
+  AutoLockHelperThreadState lock;
+  WaitForOffThreadParses(rt, lock);
+}
+
+void js::CancelOffThreadParses(JSRuntime* rt) {
+  AutoLockHelperThreadState lock;
+
+#ifdef DEBUG
+  for (const auto& task : HelperThreadState().parseWaitingOnGC(lock)) {
+    MOZ_ASSERT(!task->runtimeMatches(rt));
+  }
+#endif
+
+  // Instead of forcibly canceling pending parse tasks, just wait for all
+  // scheduled and in progress ones to complete. Otherwise the final GC may not
+  // collect everything due to zones being used off thread.
+  WaitForOffThreadParses(rt, lock);
 
   // Clean up any parse tasks which haven't been finished by the main thread.
   auto& finished = HelperThreadState().parseFinishedList(lock);
@@ -945,9 +953,7 @@
   }
 
 #ifdef DEBUG
-  GlobalHelperThreadState::ParseTaskVector& worklist =
-      HelperThreadState().parseWorklist(lock);
-  for (const auto& task : worklist) {
+  for (const auto& task : HelperThreadState().parseWorklist(lock)) {
     MOZ_ASSERT(!task->runtimeMatches(rt));
   }
 #endif